packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts

```diff
@@ -169,22 +170,31 @@ export abstract class BaseBlockDispatcher<Q extends IQueue, DS, B> implements IB
   // Is called directly after a block is processed
   @mainThreadOnly()
   protected async postProcessBlock(height: number, processBlockResponse: ProcessBlockResponse): Promise<void> {
-    const {blockHash, dynamicDsCreated, reindexBlockHeight} = processBlockResponse;
+    const {blockHash, dynamicDsCreated, reindexBlockHeight: processReindexBlockHeight} = processBlockResponse;
+    // Rewind height received from admin api have higher priority than processed reindexBlockHeight
+    const reindexBlockHeight = this._pendingRewindHeight ?? processReindexBlockHeight;
     this.monitorService?.write(`Finished block ${height}`);
     if (reindexBlockHeight !== null && reindexBlockHeight !== undefined) {
-      if (this.nodeConfig.proofOfIndex) {
-        await this.poiSyncService.stopSync();
-        this.poiSyncService.clear();
-        this.monitorService?.write(`poiSyncService stopped, cache cleared`);
+      try {
+        if (this.nodeConfig.proofOfIndex) {
+          await this.poiSyncService.stopSync();
+          this.poiSyncService.clear();
+          this.monitorService?.write(`poiSyncService stopped, cache cleared`);
+        }
+        this.monitorService?.createBlockFork(reindexBlockHeight);
+        this.resetPendingRewindHeight();
+        await this.rewind(reindexBlockHeight);
+        this.setLatestProcessedHeight(reindexBlockHeight);
+        // Bring poi sync service back to sync again.
+        if (this.nodeConfig.proofOfIndex) {
+          void this.poiSyncService.syncPoi();
+        }
+        this.eventEmitter.emit(IndexerEvent.RewindSuccess, {success: true, height: reindexBlockHeight});
+        return;
+      } catch (e) {
+        this.eventEmitter.emit(IndexerEvent.RewindFailure, {success: false, message: e});
```

Comment:
It would be good to retain the stack trace though

```suggestion
        this.eventEmitter.emit(IndexerEvent.RewindFailure, {success: false, message: e.message});
```